glad2 Rust Multi-Context

solarane opened the pull request:
Hopefully fixes #254.

It's 6 AM right now. I worked on this through the night. Probably not the best idea.
Some things to note:
- I only performed basic sanity checks; integration tests are needed
- ~~Somehow, I ended up changing the order of certain aliases even without the `--mx` flag. I'm 99% sure I just changed the order but again, integration tests are needed~~
- The formatting of `impl.rs` is a bit weird. It was a bit weird before I laid my hands on it, and it is probably weirder now because I've never used `jinja2` before. Some code review would probably be good.
- ~~Currently, with `--mx`, I don't cache function pointers. This means that for every function with an alias, for every alias, a new function pointer is created (although, of course, they are then immediately dropped). This should probably be addressed.~~

Now I'm going to go get some sleep. If y'all see something wrong, please let me know.

(CC: @Dav1dde)

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -1,5 +1,5 @@
 {% import 'template_utils.rs' as template_utils with context %}
-
+{{ '#![allow(non_snake_case)]' if options.mx }}
```
Just allows allow, makes the template more readable and doesn't matter in the generated code

solarane replied:
Makes sense :) Should be fixed in https://github.com/Dav1dde/glad/pull/256/commits/3c1d7934900137cecf580409a64b890e30e97a1c.

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -54,20 +67,31 @@ pub mod enumerations {
 
 pub mod functions {
-    #![allow(non_snake_case, unused_variables, dead_code)]
+    #![allow({{ 'non_snake_case, ' if not options.mx }}unused_variables, dead_code)]
```
Same here and all the other's just allow what you have to allow

solarane replied:
Should be fixed in https://github.com/Dav1dde/glad/pull/256/commits/3c1d7934900137cecf580409a64b890e30e97a1c.

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -54,20 +67,31 @@ pub mod enumerations {
+    {% if options.mx %}
     {% for command in feature_set.commands %}
-    {{ template_utils.protect(command) }} #[inline] pub unsafe fn {{ command.name|no_prefix }}({{ command|params }}) -> {{ command.proto.ret|type }} { mem::transmute::<_, extern "system" fn({{ command|params('types') }}) -> {{ command.proto.ret|type }}>(storage::{{ command.name|no_prefix }}.ptr)({{ command|params('names') }}) }
+    {{ '    ' }}{{ template_utils.protect(command) }} pub _{{ command.name|no_prefix }}: FnPtr,
```
`{{ '    ' }}` can just indent by that much ;)

solarane replied:
Heh, cool! Didn't know that :D Fixed.

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -76,8 +100,16 @@ mod storage {
+{% endif %}
 
-pub fn load<F>(mut loadfn: F) where F: FnMut(&'static str) -> *const raw::c_void {
+pub fn load<F>(mut loadfn: F) {{ '-> functions::Gl' if options.mx }} where F: FnMut(&'static str) -> *const raw::c_void {
```
I feel like having to separate loader functions would be more readable:

```
if mx:
  define load() -> GL
else:
  define load()
```

solarane replied:
I assume you mean just the function signature? If so, fixed in https://github.com/Dav1dde/glad/pull/256/commits/3c1d7934900137cecf580409a64b890e30e97a1c. Think it's definitely more readable now.

Dav1dde replied:
Yeah the function signature or even the whole function, since you're basically splitting the body into two separate parts anyways.

Dav1dde commented:
Looks like I submitted my review the same minute you updated the branch. Just ignore what's outdated, it's just style stuff anyways.

solarane commented:
> Currently, with `--mx`, I don't cache function pointers. This means that for every function with an alias, for every alias, a new function pointer is created (although, of course, they are then immediately dropped). This should probably be addressed.

Fixed. Bindings generated with `--mx` is still 10 lines shorter than those generated without it.

> Somehow, I ended up changing the order of certain aliases even without the `--mx` flag. I'm 99% sure I just changed the order but again, integration tests are needed

Upon closer inspection, it actually seems that the order of certain aliases is nondeterministic. Not sure if this is intended behavior or a bug.

Anyways, both issues with the original commit are now fixed.

Dav1dde commented:
> > Somehow, I ended up changing the order of certain aliases even without the `--mx` flag. I'm 99% sure I just changed the order but again, integration tests are needed
> 
> Upon closer inspection, it actually seems that the order of certain aliases is nondeterministic. Not sure if this is intended behavior or a bug.

A "bug", looks like the [function generating the aliases](https://github.com/Dav1dde/glad/blob/glad2/glad/generator/util.py#L48) is returning a dict, should probably be some kind of sorted/ordered dictionary.

Can ignore this for now, I'll fix this separately.

solarane commented:
> Can ignore this for now, I'll fix this separately.
Gotcha.

Integration tests were failing (non-`mx`) because of a small oversight (see https://github.com/Dav1dde/glad/pull/256/commits/3c1d7934900137cecf580409a64b890e30e97a1c), now fixed. Style things you mentioned should also be fixed. Thanks for the quick feedback!

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -76,7 +98,25 @@ mod storage {
 
+{% if options.mx %}
+pub fn load<F>(mut loadfn: F) -> functions::Gl where F: FnMut(&'static str) -> *const raw::c_void {
+    let mut gl = Gl {
```
doesn't need to be mut

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -1,5 +1,5 @@
 {% import 'template_utils.rs' as template_utils with context %}
-
+#![allow(non_snake_case)]
```
From my tests this is not necessary, and if it was we should move it to the block where it is required (as with the other allows)

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -54,20 +65,31 @@ pub mod enumerations {
+        {% for command in feature_set.commands %}
+        {{ template_utils.protect(command) }} #[inline] pub unsafe fn {{ command.name|no_prefix }}({{ '&self, ' if options.mx }}{{ command|params }}) -> {{ command.proto.ret|type }} { mem::transmute::<_, extern "system" fn({{ command|params('types') }}) -> {{ command.proto.ret|type }}>({{ 'self._' if options.mx else 'storage::' }}{{ command.name|no_prefix }}.ptr)({{ command|params('names') }}) }
+        {% endfor %}
+    {{ '}' if options.mx }}
```
```rust
    {% if options.mx %}
    pub struct Gl {
        {% for command in feature_set.commands %}
        {{ template_utils.protect(command) }} pub(super) _{{ command.name|no_prefix }}: FnPtr,
        {% endfor %}
    }
    impl Gl {
    {% endif %}

    {% for command in feature_set.commands %}
    {{ template_utils.protect(command) }} #[inline] pub unsafe fn {{ command.name|no_prefix }}({{ '&self, ' if options.mx }}{{ command|params }}) -> {{ command.proto.ret|type }} { mem::transmute::<_, extern "system" fn({{ command|params('types') }}) -> {{ command.proto.ret|type }}>({{ 'self._' if options.mx else 'storage::' }}{{ command.name|no_prefix }}.ptr)({{ command|params('names') }}) }
    {% endfor %}

    {{ '}' if options.mx }}
```

An idea of making it a bit more readble, clearly visible which block is the `if options.mx` block and one less template if

solarane replied:
Indents will be wrong this way, but shouldn't be an issue. definitely much more readable

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -54,20 +65,31 @@ pub mod enumerations {
 
+{% if not options.mx %}
 mod storage {
-    #![allow(non_snake_case, non_upper_case_globals)]
```
Still need this if you remove the global allow

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -12,6 +12,16 @@ pub struct FnPtr {
 
 impl FnPtr {
+    {% if options.mx %}
+    pub fn new<F>(loadfn: &mut F, name: &'static str) -> FnPtr where F: FnMut(&'static str) -> *const raw::c_void {
```
What about passing the `raw::c_void` as argument instead of the loader and the name (the loading logic shouldn't be associated with the `FnPtr` imo).

```rust
FnPtr::new(load("glClear"))
```

Dav1dde commented:
Thanks for the work! I just noted a few minor things rest looks good.

Also reminded me that I have to give Rust a little bit more love, clean up the visibility (there is way too much pub) and use `FnPtr::empty()` instead of initializing manually on every function...

Dav1dde commented:
@solarane if you want I can take care of the last minor changes

solarane commented:
@Dav1dde oh god, so sorry. I initially had this on my list of TODOs for later, but it completely slipped out of my mind D: I will get to it tomorrow (it's 2 AM where I am currently) and resolve your reviews. Sorry again.

Dav1dde commented:
No worries take your time

solarane commented:
@Dav1dde All your reviews should now be addressed. Tests are passing. Ready for merge?

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -12,6 +11,15 @@ pub struct FnPtr {
 
 impl FnPtr {
+    {% if options.mx %}
+    pub fn new<F>(loaded: *const raw::c_void) -> FnPtr where F: FnMut(&'static str) -> *const raw::c_void {
```
`F` is not used anymore and leads to compiler errors "cannot infer F"

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -76,7 +97,25 @@ mod storage {
+
+{% if options.mx %}
+pub fn load<F>(mut loadfn: F) -> functions::Gl where F: FnMut(&'static str) -> *const raw::c_void {
+    let gl = Gl {
```
I told you this doesn't need to be mut, sorry in fact it has to be `mut` because of the later `aliased` calls, but we need to silence the warning in case there are no aliased functions

Dav1dde commented on glad/generator/rust/templates/impl.rs:
```diff
@@ -26,7 +34,8 @@ impl FnPtr {
-
+    
+    {% endif %}
     pub fn aliased(&mut self, other: &FnPtr) {
```
It is possible that there are no aliased functions, then this is dead code, the struct or function needs `#[allow(dead_code)]`

Dav1dde commented:
@solarane I am testing the review feature (never really used it before), I think you get that changes are requested because I have to accept the first?

Anyways there are a few things needed in order for it to compile, I pushed an example (and simple test) to the [rust-mx-add](https://github.com/Dav1dde/glad/tree/rust-mx-add) branch. If you're using it just make sure to generate once without extensions (default) and with all extensions (for no aliasing and aliasing).

solarane commented:
@Dav1dde All right, all should be fixed.

In addition to stuff you reviewed, I saw that you changed `pub struct FnPtr` to `pub(crate)`. Since it's at the root of the crate, I'm pretty sure that's just equivalent to not making it `pub`, which is what I did. Also this time, I tested that everything compiles and runs as expected. :3

Dav1dde commented:
@solarane thank you for your work! I went ahead and rebased the commits and pushed them as 91c7599

solarane commented:
Thanks for the merge!
~~I didn't get the contributor badge, which is sad. Guess that's because this PR wasn't merged directly? D:~~
I've also noticed a bug (?) with how multiple contexts interact with multiple threads, which would require changing the internal representations of the function pointers. I'll open a new issue on that, and probably a PR sometime soon...

Dav1dde commented:
@solarane yeah I wish github wasn't like this, all my commits to glad2 do not count towards the statistics either.

Please do open an issue.
